### PR TITLE
Fix coloring flag not working on GUI with granular strategies

### PR DESCRIPTION
**Pull request by janko**
The `test#base#no_colors()` function was looking at the global strategy variable, which can be a dictionary when per-granularity strategies was specified. This function doesn't know the granularity, since the `build_args()` runner function is called from `test#execute()`, which doesn't have granularity knowledge (since that function can be called directly via runner commands such as `:RSpec`).

Instead of requiring runners to call into our function for determining whether coloring can be used, we change the code to determine this before calling `build_args()`, and then just pass the boolean value into the function. At that point the strategy has already been resolved, so we can just use it right there and then.

I didn't find a way in VimScript to check how many arguments a function supports, so we had to resort to try/catch for backwards compatibility. We could have updated all runners to accept the second argument, but that wouldn't solve any custom-made runners that people might have created locally.

Closes #429

### CODE REVIEW

**Comment by janko** [comment]
@willfish This is what I had in mind ☝️ (though the `catch` blog doesn't work on CI at the moment)

File: autoload/test/base.vim [hunk 1]
```diff
@@ -30,8 +30,14 @@ function! test#base#executable(runner) abort
   endif
 endfunction
 
-function! test#base#build_args(runner, args) abort
-  return test#{a:runner}#build_args(a:args)
+function! test#base#build_args(runner, args, strategy) abort
```
Comment (willfish) [inline autoload/test/base.vim]:
```suggestion
function! test#base#build_args(runner, args, strategy)
```

Not super familiar with vim script, but do we need to kill the abort if we mean to use exceptions to control the branches of the function?

**Comment by willfish** [comment]
This approach seems more robust to me, yeah 🎉

File: autoload/test/base.vim [hunk 2]
```diff
@@ -30,8 +30,14 @@ function! test#base#executable(runner) abort
...
+  let no_color = has('gui_running') && strategy ==# 'basic'
```
Comment (willfish) [inline autoload/test/base.vim]:
```suggestion
  let no_color = has('gui_running') && a:strategy ==# 'basic'
```

**Comment by willfish** [comment]
With those changes it's working for me using MacVim and Dispatch and the multiple strategy config.

```vim
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
" => vim-test
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

let test#strategy = {
      \ 'nearest': 'dispatch',
      \ 'file':    'dispatch',
      \ 'suite':   'dispatch_background',
      \}

let test#ruby#rspec#executable = 'bundle exec rspec'

nnoremap <silent> <Leader>x :TestNearest<CR>
nnoremap <silent> <Leader>t :TestFile<CR>
nnoremap <silent> <Leader>r :TestSuite<CR>
nnoremap <silent> <Leader>e :TestLast<CR>
nnoremap <silent> <Leader>l :TestVisit<CR>
```

I tried nearest, file and suite and all work

image

**Comment by willfish** [comment]
@janko Do you mind reviewing this again?

**Comment by janko** [comment]
@willfish I appreciate the code review 🙏

I realized that I don't have time to maintain this project anymore, so I've moved the repo to a new `vim-test` organization, and have added you as a member (you should have push access).

You're free to get this pull request merged and do any other changes you want in the future 😉

**Comment by willfish** [comment]
@janko The specs were failing on older versions of vim. I've upgraded and cached the one we build in travis. It doesn't add loads of time to the build and makes these specs pass. Let me know if you think this is okay and I'll merge this change :)

**Comment by janko** [comment]
Looks great to me! 👍 And I really appreciate you picking this up ❤️

**Comment by willfish** [comment]
No worries :).

**Pull request merged**